keps/sig-node/3721-support-for-env-files/README.md

```diff
@@ -0,0 +1,559 @@
+file as it can contain sensitive information.
+
+Below are the ones we mapped and their outcome once this KEP is implemented.
+
+|Scenario| API Server Result | Kubelet Result |
+|--------|-------------------|----------------|
+|1. The volumeName specified in `FileEnvSource` or `FileKeySelector` field does not exist | Pod creation fails with an error |  |
+|2. The volumeName specified in `FileEnvSource` or `FileKeySelector` is not an emptyDir | Pod creation fails with an error |  |
+|3. The filepath specified in `FileEnvSource` field does not exist | Pod created | Container fails to start and error message in event.|
+|4. The filepath specified in `FileEnvSource` field does not exist but `optional` field is set to true. | Pod created | Container starts and env vars are not populated.|
+|5. Either the filepath or key specified in `FileKeySelector` field does not exist | Pod created | Container fails to start and error message in event.|
+|6. Either the filepath or key specified in `FileKeySelector` field exist but `optional` field is set to true | Pod created | Container starts and env vars are not populated. |
+|7. The specified file is not a parsable env file. | Pod created | Container fails to start and error message is reported in the events.|
+|8. The specified file contains invalid env var names. | Pod created | Container fails to start and erorr message is reported in the events.|
+
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+No
+
+##### Unit tests
+
+- [Pod Validation
+  tests](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation_test.go)
+- TODO
+
+##### Integration tests
+
+- Pods tests: https://github.com/kubernetes/kubernetes/blob/91ee30074bee617d52fc24dc85132fe948aa5153/test/integration/pods/pods_test.go
+
+##### e2e tests
+
+The tests will be guarded by the `[Feature:EnvFiles]` tag.
+- Pods e2e test: https://github.com/kubernetes/kubernetes/blob/91ee30074bee617d52fc24dc85132fe948aa5153/test/e2e/common/node/pods.go
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature flag.
+- unit and integration tests completed.
+- Initial e2e tests completed and enabled.
+
+#### Beta
+
+- Feature gate is enabled by default.
+- Allowing time for feedback.
+
+#### GA
+
+- TBD
+
+### Upgrade / Downgrade Strategy
+
+kube-apiserver should be updated before kubelets in that order. Upgrade involves
+draining all pods from the node, update to a matching kubelet and making the
+node schedulable again. Downgrade involves doing the above in reverse.
+
+### Version Skew Strategy
+
+Since kubelet versions must not be ahead of kube-apiserver versions, the older
+versions of kubelet or if the feature gate is disabled in kubelet, the fields
+will be ignored.
```

Comment:
For my understanding, what of the Optional field is set to true in a higher version and the lower version doesnt have the feature enabled- how is that handled?

Reply:
`kube-apiserver` at a higher version with this feature enabled will accept the request with new API fields. But `kubelet` at a lower version with this feature disabled will ignore this new API fields and start the workload without setting the env vars from this feature.